snorkel/candidates.py

```diff
@@ -31,72 +28,86 @@ class CandidateExtractor(UDF):
+
+    class CandidateExtractorUDF(UDF):
+        def __init__(self, candidate_class, cspaces, matchers, self_relations, nested_relations, symmetric_relations, in_queue):
+            self.candidate_class     = candidate_class
+            self.candidate_spaces    = cspaces if type(cspaces) in [list, tuple] else [cspaces]
+            self.matchers            = matchers if type(matchers) in [list, tuple] else [matchers]
+            self.nested_relations    = nested_relations
+            self.self_relations      = self_relations
+            self.symmetric_relations = symmetric_relations
+
+            # Check that arity is same
+            if len(self.candidate_spaces) != len(self.matchers):
+                raise ValueError("Mismatched arity of candidate space and matcher.")
+            else:
+                self.arity = len(self.candidate_spaces)
+
+            # Make sure the candidate spaces are different so generators aren't expended!
+            self.candidate_spaces = map(deepcopy, self.candidate_spaces)
+
+            # Preallocates internal data structures
+            self.child_context_sets = [None] * self.arity
+            for i in range(self.arity):
+                self.child_context_sets[i] = set()
+
+            super(CandidateExtractor.CandidateExtractorUDF, self).__init__(in_queue=in_queue)
+
+        def apply(self, context, split=0, clear=True):
+            # Generate TemporaryContexts that are children of the context using the candidate_space and filtered
+            # by the Matcher
+            for i in range(self.arity):
+                self.child_context_sets[i].clear()
+                for tc in self.matchers[i].apply(self.candidate_spaces[i].apply(context)):
+                    tc.load_id_or_insert(self.session)
+                    self.child_context_sets[i].add(tc)
+
+            # Generates and persists candidates
+            candidate_args = {'split': split}
+            for args in product(*[enumerate(child_contexts) for child_contexts in self.child_context_sets]):
+
+                # TODO: Make this work for higher-order relations
+                if self.arity == 2:
+                    ai, a = args[0]
+                    bi, b = args[1]
+
+                    # Check for self-joins, "nested" joins (joins from span to its subspan), and flipped duplicate
+                    # "symmetric" relations
+                    if not self.self_relations and a == b:
+                        continue
+                    elif not self.nested_relations and (a in b or b in a):
+                        continue
+                    elif not self.symmetric_relations and ai > bi:
+                        continue
+
+                # Assemble candidate arguments
+                for i, arg_name in enumerate(self.candidate_class.__argnames__):
+                    candidate_args[arg_name + '_id'] = args[i][1].id
+
+                # Checking for existence
+                if not clear:
```

Comment:
We should automatically set `clear=True` if we know we're deleting all the candidates before running, right?